File: app/src/main/java/pl/karol202/paintplus/tool/shape/circle/ShapeCircle.java
package pl.karol202.paintplus.tool.shape.circle;

import android.graphics.*;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.image.Image;
import pl.karol202.paintplus.tool.shape.OnShapeEditListener;
import pl.karol202.paintplus.tool.shape.Shape;
import pl.karol202.paintplus.tool.shape.ShapeProperties;
import pl.karol202.paintplus.util.Utils;

public class ShapeCircle extends Shape
{
	private boolean fill;
	private int circleWidth;
	
	private boolean circleCreated;
	private Point center;
	private float radius;
	
	private Point draggingStart;
	private int draggedIndex;
	private Point centerAtBeginning;
	private float radiusAtBeginning;
	
	public ShapeCircle(Image image, OnShapeEditListener shapeEditListener)
	{
		super(image, shapeEditListener);
		this.fill = false;
		this.circleWidth = 30;
		
		update();
	}
	
	@Override
	public int getName()
	{
		return R.string.shape_circle;
	}
	
	@Override
	public int getIcon()
	{
		return R.drawable.ic_shape_circle_black_24dp;
	}
	
	@Override
	public Class<? extends ShapeProperties> getPropertiesClass()
	{
		return CircleProperties.class;
	}
	
	@Override
	public void onTouchStart(int x, int y)
	{
		if(!isInEditMode()) enableEditMode();
		if(!circleCreated) setCenterPoint(new Point(x, y));
		else
		{
			float distanceToCenter = calcDistance(center, x, y);
			float distanceToRadius = Math.abs(distanceToCenter - radius);
			
			draggingStart = new Point(x, y);
			centerAtBeginning = center;
			radiusAtBeginning = radius;
			
			if(Math.min(distanceToCenter, distanceToRadius) > getMaxTouchDistance()) draggedIndex = -1;
			else if(distanceToCenter < distanceToRadius) draggedIndex = 0;
			else draggedIndex = 1;
		}
	}
	
	@Override
	public void onTouchMove(int x, int y)
	{
		Point current = new Point(x, y);
		if(!circleCreated) dragRadius(current);
		else drag(current);
	}
	
	@Override
	public void onTouchStop(int x, int y)
	{
		onTouchMove(x, y);
		
		circleCreated = true;
		draggingStart = null;
	}
	
	private void drag(Point current)
	{
		if(draggedIndex == 0) dragCenter(current);
		else if(draggedIndex == 1) dragRadius(current);
	}
	
	private void dragCenter(Point current)
	{
		Point delta = new Point(current);
		delta.x -= draggingStart.x;
		delta.y -= draggingStart.y;
		
		Point newCenter = new Point(centerAtBeginning);
		newCenter.offset(delta.x, delta.y);
		setCenterPoint(newCenter);
	}
	
	private void dragRadius(Point current)
	{
		if(draggingStart != null)
		{
			double theta = Math.toRadians(Utils.getAngle(center, current));
			float rCurrent = calcDistance(center, current.x, current.y);
			float rDraggingStart = calcDistance(center, draggingStart.x, draggingStart.y);
			float rBeginning = radiusAtBeginning;
			
			float rDelta = rCurrent - rDraggingStart;
			float rResult = rBeginning + rDelta;
			
			float x = (float) (rResult * Math.cos(theta)) + center.x;
			float y = (float) (rResult * Math.sin(theta)) + center.y;
			PointF result = new PointF(x, y);
			getHelpersManager().snapPoint(result);
			
			radius = calcDistance(center, (int) result.x, (int) result.y);
		}
		else
		{
			PointF snapped = new PointF(current);
			getHelpersManager().snapPoint(snapped);
			radius = calcDistance(center, (int) snapped.x, (int) snapped.y);
		}
	}
	
	private void setCenterPoint(Point point)
	{
		PointF snapped = new PointF(point);
		getHelpersManager().snapPoint(snapped);
		center = new Point((int) snapped.x, (int) snapped.y);
	}
	
	@Override
	public void expandDirtyRect(Rect dirtyRect)
	{
		dirtyRect.left = Math.min(dirtyRect.left, (int) (center.x - radius - circleWidth));
		dirtyRect.top = Math.min(dirtyRect.top, (int) (center.y - radius - circleWidth));
		dirtyRect.right = Math.max(dirtyRect.right, (int) (center.x + radius + circleWidth));
		dirtyRect.bottom = Math.max(dirtyRect.bottom, (int) (center.y + radius + circleWidth));
	}
	
	@Override
	public void onScreenDraw(Canvas canvas, boolean translucent)
	{
		if(center == null || radius == -1) return;
		updateColor(translucent);
		canvas.drawCircle(center.x, center.y, radius, getPaint());
	}
	
	@Override
	public Rect getBoundsOfShape()
	{
		int halfSize = (int) radius + circleWidth;
		return new Rect(center.x - halfSize, center.y - halfSize,
					   center.x + halfSize, center.y + halfSize);
	}
	
	@Override
	public void apply(Canvas imageCanvas)
	{
		if(center == null || radius == -1) return;
		update();
		imageCanvas.drawCircle(center.x, center.y, radius, getPaint());
		cleanUp();
	}
	
	@Override
	public void cancel()
	{
		cleanUp();
	}
	
	@Override
	public void update()
	{
		getPaint().setStyle(fill ? Paint.Style.FILL_AND_STROKE : Paint.Style.STROKE);
		getPaint().setStrokeWidth(circleWidth);
		super.update();
	}
	
	@Override
	public void cleanUp()
	{
		circleCreated = false;
		center = null;
		radius = -1;
		super.cleanUp();
	}
	
	@Override
	public void enableEditMode()
	{
		circleCreated = false;
		center = null;
		radius = -1;
		super.enableEditMode();
	}
	
	boolean isFilled()
	{
		return fill;
	}
	
	void setFill(boolean fill)
	{
		this.fill = fill;
		update();
	}
	
	int getCircleWidth()
	{
		return circleWidth;
	}
	
	void setCircleWidth(int circleWidth)
	{
		this.circleWidth = circleWidth;
		update();
	}
}

File: app/src/main/java/pl/karol202/paintplus/tool/shape/circle/ShapeCircle.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.shape.circle

import android.content.Context
import android.graphics.*
import androidx.core.graphics.minus
import androidx.core.graphics.plus
import androidx.core.graphics.toPoint
import androidx.core.graphics.toPointF
import pl.karol202.paintplus.R
import pl.karol202.paintplus.helpers.HelpersService
import pl.karol202.paintplus.image.ColorsService
import pl.karol202.paintplus.image.EffectsService
import pl.karol202.paintplus.image.ViewService
import pl.karol202.paintplus.tool.shape.AbstractShape
import pl.karol202.paintplus.util.MathUtils
import pl.karol202.paintplus.util.MathUtils.distance
import pl.karol202.paintplus.util.cache
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin

class ShapeCircle(context: Context,
                  viewService: ViewService,
                  colorsService: ColorsService,
                  effectsService: EffectsService,
                  private val helpersService: HelpersService) :
		AbstractShape(context, viewService, colorsService, effectsService)
{
	private enum class DragType
	{
		CENTER, RADIUS
	}

	private sealed class State
	{
		object Idle : State()

		sealed class Edit : State()
		{
			data class Initial(override val center: Point,
			                   override val radius: Float) : Edit()

			data class CenterDrag(override val center: Point,
			                      override val radius: Float,
			                      val dragStart: Point,
			                      val centerAtStart: Point) : Edit()

			data class RadiusDrag(override val center: Point,
			                      override val radius: Float,
			                      val dragStart: Point,
			                      val radiusAtStart: Float) : Edit()

			abstract val center: Point
			abstract val radius: Float
		}
	}

	override val name get() = R.string.shape_circle
	override val icon get() = R.drawable.ic_shape_circle_black_24dp
	override val propertiesClass get() = CircleProperties::class.java

	var isFilled by notifying(false)
	var circleWidth by notifying(30)

	override val standardPaint by cache({super.standardPaint}, {isFilled}, {circleWidth}) { standardPaint, fill, width ->
		Paint(standardPaint).apply {
			style = if(fill) Paint.Style.FILL_AND_STROKE else Paint.Style.STROKE
			strokeWidth = width.toFloat()
		}
	}
	override val translucentPaint by cache({super.translucentPaint}, {isFilled}, {circleWidth}) { translucentPaint, fill, width ->
		Paint(translucentPaint).apply {
			style = if(fill) Paint.Style.FILL_AND_STROKE else Paint.Style.STROKE
			strokeWidth = width.toFloat()
		}
	}

	private var state: State by notifying(State.Idle)

	override val isInEditMode get() = state is State.Edit
	override val bounds get() = when(val state = state)
	{
		is State.Idle -> null
		is State.Edit ->
		{
			val halfSize = state.radius.toInt() + circleWidth
			Rect(state.center.x - halfSize, state.center.y - halfSize,
			     state.center.x + halfSize, state.center.y + halfSize)
		}
	}

	override fun onTouchStart(point: Point)
	{
		state = getInitialState(point)
	}

	private fun getInitialState(point: Point) = when(val state = state)
	{
		is State.Idle -> State.Edit.Initial(snapPoint(point), 0f)
		is State.Edit ->
		{
			when(getDragType(point, state))
			{
				null -> State.Edit.Initial(state.center, state.radius)
				DragType.CENTER -> State.Edit.CenterDrag(state.center, state.radius, point, state.center)
				DragType.RADIUS -> State.Edit.RadiusDrag(state.center, state.radius, point, state.radius)
			}

		}
	}

	private fun getDragType(point: Point, state: State.Edit): DragType?
	{
		val distanceToCenter: Float = distance(state.center, point)
		val distanceToRadius = abs(distanceToCenter - state.radius)
		return when
		{
			min(distanceToCenter, distanceToRadius) > maxTouchDistancePx -> null
			distanceToCenter < distanceToRadius -> DragType.CENTER
			else -> DragType.RADIUS
		}
	}

	override fun onTouchMove(point: Point)
	{
		state = getEditedState(point) ?: return
	}

	override fun onTouchStop(point: Point)
	{
		state = getEditedState(point) ?: return
	}

	private fun getEditedState(point: Point) = when(val state = state)
	{
		is State.Idle -> null
		is State.Edit.Initial -> state.copy(radius = distance(state.center, snapPoint(point)))
		is State.Edit.CenterDrag ->
		{
			val newPoint = state.centerAtStart + (point - state.dragStart)
			state.copy(center = snapPoint(newPoint))
		}
		is State.Edit.RadiusDrag ->
		{
			val theta = Math.toRadians(MathUtils.getAngle(state.center, point))
			val rCurrent = distance(state.center, point)
			val rDraggingStart = distance(state.center, state.dragStart)
			val rBeginning = state.radiusAtStart
			val rDelta = rCurrent - rDraggingStart
			val rResult = rBeginning + rDelta
			val x = (rResult * cos(theta)).toFloat() + state.center.x
			val y = (rResult * sin(theta)).toFloat() + state.center.y
			val result = PointF(x, y)

			state.copy(radius = distance(state.center.toPointF(), snapPoint(result)))
		}
	}

	private fun snapPoint(point: Point) = helpersService.snapPoint(point.toPointF()).toPoint()

	private fun snapPoint(point: PointF) = helpersService.snapPoint(point)

	override fun drawOnLayer(canvas: Canvas, translucent: Boolean)
	{
		val editState = state as? State.Edit ?: return
		val paint = if(translucent) translucentPaint else standardPaint
		canvas.drawCircle(editState.center.x.toFloat(), editState.center.y.toFloat(), editState.radius, paint)
	}

	override fun apply(imageCanvas: Canvas)
	{
		val editState = state as? State.Edit ?: return
		imageCanvas.drawCircle(editState.center.x.toFloat(), editState.center.y.toFloat(), editState.radius, standardPaint)
		cancel()
	}

	override fun cancel()
	{
		state = State.Idle
	}
}
